adapters/adgeneration/adgeneration.go

```diff
@@ -0,0 +1,264 @@
+		"&sdkname=prebidserver" +
+		"&adapterver=" + adg.version
+	if getSizes(&imp) != "" {
+		endpoint += "&sizes=" + getSizes(&imp)
+	}
+	if request.Site != nil && request.Site.Page != "" {
+		endpoint += "&tp=" + request.Site.Page
+	}
+
+	return endpoint, nil
+}
+
+func unmarshalExtImpBidder(imp *openrtb.Imp) (*adapters.ExtImpBidder, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, err
+	}
+	return &bidderExt, nil
+}
+
+func unmarshalExtImpAdgeneration(ext *adapters.ExtImpBidder) (*openrtb_ext.ExtImpAdgeneration, error) {
+	var adgExt openrtb_ext.ExtImpAdgeneration
+	if err := json.Unmarshal(ext.Bidder, &adgExt); err != nil {
+		return nil, err
+	}
+	if adgExt.Id == "" {
+		return nil, errors.New("No Location ID in ExtImpAdgeneration.")
+	}
+	return &adgExt, nil
+}
+
+func getSizes(imp *openrtb.Imp) string {
+	if imp.Banner == nil || len(imp.Banner.Format) == 0 {
+		return ""
+	}
+	var sizeStr string
+	for _, v := range imp.Banner.Format {
+		sizeStr += strconv.FormatUint(v.W, 10) + "×" + strconv.FormatUint(v.H, 10) + ","
+	}
+	if len(sizeStr) > 0 && strings.LastIndex(sizeStr, ",") == len(sizeStr)-1 {
+		sizeStr = sizeStr[:len(sizeStr)-1]
+	}
+	return sizeStr
+}
+
+func (adg *AdgenerationAdapter) getCurrency(request *openrtb.BidRequest) string {
+	if len(request.Cur) <= 0 {
+		return adg.defaultCurrency
+	} else {
+		for _, c := range request.Cur {
+			if adg.defaultCurrency == c {
+				return c
+			}
+		}
+		return request.Cur[0]
+	}
+}
+
+func (adg *AdgenerationAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+	var bidResp adgServerResponse
+	err := json.Unmarshal(response.Body, &bidResp)
+	if err != nil {
+		return nil, []error{err}
+	}
+	if len(bidResp.Results) <= 0 {
+		return nil, nil
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+	var impId string
+	var bitType openrtb_ext.BidType
+	var adm string
+	for _, v := range internalRequest.Imp {
+		bidderExt, err := unmarshalExtImpBidder(&v)
+		if err != nil {
+			return nil, []error{&errortypes.BadServerResponse{
+				Message: err.Error(),
+			},
+			}
+		}
+		adgExt, err := unmarshalExtImpAdgeneration(bidderExt)
+		if err != nil {
+			return nil, []error{&errortypes.BadServerResponse{
+				Message: err.Error(),
+			},
+			}
+		}
```

Comment:
These lines are almost identical as `getRequestUri` lines 77 to 89. Can we put this functionality in a separate function so both `getRequestUri` and `MakeBids` call it?
```
 76 func (adg *AdgenerationAdapter) getRequestUri(request *openrtb.BidRequest, index int) (string, error) {
 77     imp := request.Imp[index]
 78     bidderExt, err := adg.unmarshalExtImpBidder(&imp)
 79     if err != nil {
 80         return "", &errortypes.BadInput{
 81             Message: err.Error(),
 82         }
 83     }
 84     adgExt, err := adg.unmarshalExtImpAdgeneration(&bidderExt)
 85     if err != nil {
 86         return "", &errortypes.BadInput{
 87             Message: err.Error(),
 88         }
 89     }
adapters/adgeneration/adgeneration.go
```